backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt

```diff
@@ -154,40 +162,192 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
+        it.putValueArgument(0, readMemory)
+    }
+}
+
+private fun InteropCallContext.readPointed(nativePtr: IrExpression): IrExpression {
+    return builder.irCall(symbols.interopInterpretNullablePointed).also {
+        it.putValueArgument(0, nativePtr)
+    }
+}
+
+/** Returns non-null result if [callSite] is accessor to:
+ *  1. T.value, T : CEnumVar
+ *  2. T.<field-name>, T : CStructVar and accessor is annotated with
+ *      [kotlinx.cinterop.internal.CStruct.MemberAt] or [kotlinx.cinterop.internal.CStruct.BitField]
  */
-internal fun tryGenerateEnumVarValueAccess(
+internal fun tryGenerateInteropMemberAccess(
         callSite: IrCall,
         symbols: KonanSymbols,
         builder: IrBuilderWithScope
-): IrExpression? = if (isEnumVarValueAccessor(callSite.symbol.owner, symbols))
-    generateInteropCall(symbols, builder) { generateEnumVarValueAccess(callSite) }
-else null
+): IrExpression? = when {
+    isEnumVarValueAccessor(callSite.symbol.owner, symbols) ->
+        generateInteropCall(symbols, builder) { generateEnumVarValueAccess(callSite) }
+    isMemberAtAccessor(callSite.symbol.owner) ->
+        generateInteropCall(symbols, builder) { generateMemberAtAccess(callSite) }
+    isBitFieldAccessor(callSite.symbol.owner) ->
+        generateInteropCall(symbols, builder) { generateBitFieldAccess(callSite) }
+    else -> null
+}
 
 private fun InteropCallContext.generateEnumVarValueAccess(callSite: IrCall): IrExpression {
     val accessor = callSite.symbol.owner
     val nativePtr = builder.irCall(symbols.interopNativePointedRawPtrGetter).also {
         it.dispatchReceiver = callSite.dispatchReceiver!!
     }
+    return when {
+        accessor.isGetter -> readEnumValueFromMemory(nativePtr, accessor.returnType)
+        accessor.isSetter -> writeEnumValueToMemory(nativePtr, callSite.getValueArgument(0)!!)
+        else -> error("")
+    }
+}
+
+private fun InteropCallContext.generateMemberAtAccess(callSite: IrCall): IrExpression {
+    val accessor = callSite.symbol.owner
+    val memberAt = accessor.getAnnotation(RuntimeNames.cStructMemberAt)!!
+    val offset = (memberAt.getValueArgument(0) as IrConst<Long>).value
+    val fieldPointer = calculateFieldPointer(callSite.dispatchReceiver!!, offset)
     return when {
         accessor.isGetter -> {
-            val enumClass = accessor.returnType.getClass()!!
-            val enumPrimitiveType = enumClass.properties.single { it.name.asString() == "value" }
-                    .getter!!.returnType
-            val readMemory = readPrimitiveFromMemory(nativePtr, enumPrimitiveType)
-            return convertIntegralToEnum(readMemory, accessor.returnType)
+            val type = accessor.returnType
+            when {
+                type.isCEnumType() -> readEnumValueFromMemory(fieldPointer, type)
+                type.isPrimitiveType() -> readPrimitiveFromMemory(fieldPointer, type)
+                type.isCPointer() -> readPointerFromMemory(fieldPointer)
```

Comment:
Does this work for array-typed members of structs?

Reply:
It should, but it won't work without #3826 that adds support for `CArrayPointer` typealias.

Reply:
Getting array-typed member doesn't involve reading from memory, however this branch does perform reading. How can this work?

Reply:
Ok, I see.